File: src/main/java/cz/jskrabal/proxy/pump/DataPump.java
package cz.jskrabal.proxy.pump;

import io.vertx.core.streams.ReadStream;
import io.vertx.core.streams.WriteStream;

import java.util.function.Consumer;

/**
 * Created by janskrabal on 01/07/16.
 */
public class DataPump<T> {
	private final ReadStream<T> input;
	private final WriteStream<T> output;
	private final Consumer<T> interceptor;

	public DataPump(ReadStream<T> input, WriteStream<T> output, Consumer<T> interceptor) {
		this.input = input;
		this.output = output;
		this.interceptor = interceptor;
	}

	public void start(){
		input.handler(data -> {
			interceptor.accept(data);

			output.write(data);
			if(output.writeQueueFull()) {
				input.pause();
				output.drainHandler(v -> input.resume());
			}
		});
	}
}

File: src/main/kotlin/cz/jskrabal/proxy/pump/DataPump.kt
package cz.jskrabal.proxy.pump

import io.vertx.core.streams.Pump
import io.vertx.core.streams.ReadStream
import io.vertx.core.streams.WriteStream

/**
 * Created by janskrabal on 01/07/16.
 */
class DataPump<T> private constructor(private val input: ReadStream<T>,
                                      private val output: WriteStream<T>,
                                      private inline val interceptor: (T) -> Unit) : Pump {

    private var pumped: Int = 0

    override fun start(): Pump {
        input.handler { data ->
            interceptor.invoke(data)

            output.write(data)
            if (output.writeQueueFull()) {
                input.pause()
                output.drainHandler { input.resume() }
            }
        }
        return this
    }

    override fun stop(): Pump {
        output.drainHandler(null)
        input.handler(null)
        return this
    }

    override fun setWriteQueueMaxSize(maxSize: Int): Pump {
        this.output.setWriteQueueMaxSize(maxSize)
        return this
    }

    override fun numberPumped(): Int {
        return pumped
    }

    companion object {
        fun <T> create(input: ReadStream<T>, output: WriteStream<T>): Pump {
            return Pump.pump(input, output)
        }

        fun <T> create(input: ReadStream<T>, output: WriteStream<T>, interceptor: (T) -> Unit): Pump {
            return DataPump(input, output, interceptor)
        }
    }
}


